src/integrations/GA4/utils.js

```diff
@@ -1,219 +1,325 @@
   const items = [];
-  let obj = {};
-  const contextOp = type(products) !== 'array' ? 'properties' : 'product';
-  const finalProducts = type(products) !== 'array' ? [products] : products;
-  const finalItemObj = item && type(item) === 'array' && item[0] ? item[0] : {};
-  // get the dest keys from itemParameters config
-  // append the already created item object keys (this is done to get the keys that are actually top level props in Rudder payload but GA expects them under items too)
-  finalProducts.forEach((product) => {
-    obj = {
-      ...getDestinationEventProperties(product, itemParametersConfigArray, contextOp, true),
-      ...finalItemObj,
-    };
-    items.push(obj);
-  });
+
+  // Only prepare items array if properties exists and it should have at least one key
+  if (properties && Object.keys(properties).length > 0) {
+    const item = constructPayload(properties, itemsArrayParams);
+    if (!isEmptyObject(item)) {
+      items.push(item);
+    }
+  }
+
   return items;
-}
+};
 
 /**
- * Generate ga4 page_view events payload
- * @param {*} props
+ * Returns items array for ga4 event payload
+ * @param {*} message
+ * @param {*} eventConfig
+ * @returns
  */
-function getPageViewProperty(props) {
-  return getDestinationEventProperties(props, pageEventParametersConfigArray, 'properties');
-}
+const getItemsArray = (message, eventConfig) => {
+  const { itemList, item } = eventConfig;
+  let items = [];
+  let mapRootLevelPropertiesToGA4ItemsArray = false;
+
+  if (itemList && item) {
+    items = getItemList(message);
+
+    if (!(items && items.length > 0)) {
+      mapRootLevelPropertiesToGA4ItemsArray = true;
+      items = getItem(message);
+    }
+  } else if (item) {
+    // item
+    mapRootLevelPropertiesToGA4ItemsArray = true;
+    items = getItem(message);
+  } else if (itemList) {
+    // itemList
+    items = getItemList(message);
+  }
+
+  return { items, mapRootLevelPropertiesToGA4ItemsArray };
+};
 
 /**
- * Validates weather to send userId property to GA4 or not
- * @param {*} integrations
+ * Returns ga4 standard event payload
+ * @param {*} message
+ * @param {*} eventConfig
+ * @returns
  */
-function sendUserIdToGA4(integrations) {
-  if (Object.prototype.hasOwnProperty.call(integrations, 'GA4')) {
-    const { GA4 } = integrations;
-    if (Object.prototype.hasOwnProperty.call(GA4, 'sendUserId')) {
-      return !!GA4.sendUserId;
+const prepareStandardEventParams = (message, eventConfig) => {
+  const { event, mapping } = eventConfig;
+  let payload = constructPayload(message, mapping);
+
+  // Validation for required params
+  if (Array.isArray(mapping) && mapping.length > 0) {
+    // eslint-disable-next-line no-restricted-syntax
+    for (const mappingItem of mapping) {
+      if (!payload[mappingItem.destKey] && mappingItem.required) {
+        logger.error(`Missing required value from ${JSON.stringify(mappingItem.sourceKeys)}`);
+        return null;
+      }
     }
-    return true;
   }
```

Comment:
In this block can we use every method? IMO it will be more conscise.

Reply:
some suits more as per our use case. implemented that